### PR TITLE
Load individual elements if state dict load fails

### PR DESCRIPTION
**Pull request by andrewcoh**
### Proposed change(s)

Addressing the issue of being unable to resume with GAIL/reward providers in general. If the load fails, this will copy matching elements individually and produce a lot of logger warnings as such:

```
2021-04-01 12:29:22 WARNING [torch_model_saver.py:102] Did not expect these keys ['action_model._continuous_distribution.log_sigma', 'action_model._continuous_distribution.mu.weight', 'action_model._continuous_distribution.mu.bias'] in checkpoint. Initializing
2021-04-01 12:29:22 WARNING [torch_model_saver.py:106] Failed to load for module Optimizer:value_optimizer. Initializing
2021-04-01 12:29:22 WARNING [torch_model_saver.py:98] Did not find these keys ['value_heads.value_heads.curiosity.weight', 'value_heads.value_heads.curiosity.bias', 'value_heads.value_heads.gail.weight', 'value_heads.value_heads.gail.bias'] in checkpoint. Initializing
2021-04-01 12:29:22 WARNING [torch_model_saver.py:102] Did not expect these keys ['value_heads.value_heads.extrinsic.weight', 'value_heads.value_heads.extrinsic.bias', 'value_heads.value_heads.rnd.weight', 'value_heads.value_heads.rnd.bias'] in checkpoint. Initializing
2021-04-01 12:29:22 WARNING [torch_model_saver.py:106] Failed to load for module Module:Curiosity. Initializing
2021-04-01 12:29:22 WARNING [torch_model_saver.py:106] Failed to load for module Module:GAIL. Initializing
2021-04-01 12:29:22 INFO [torch_model_saver.py:116] Resuming training from step 42896.
```

TODO:

- tests

### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/main/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/main/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/main/docs/Migrating.md) (if applicable)

### CODE REVIEW

**Comment by ervteng** [comment]
I think this warrants a doc change. In `Training-ML-Agents` under "Loading an existing model", we can say something like "If the network architecture changes, you may still load an existing model, and ML-Agents will only load the parts of the model that haven't changed. For instance, if you add a new reward signal, the existing model will load but the new reward signal will be initialized from scratch. If you have a model with a visual encoder (CNN) but change the `hidden_units`, the CNN will be loaded but the body of the network will not be."

File: ml-agents/mlagents/trainers/model_saver/torch_model_saver.py [hunk 1]
```diff
@@ -89,7 +89,25 @@ def _load_model(
         policy = cast(TorchPolicy, policy)
 
         for name, mod in modules.items():
-            mod.load_state_dict(saved_state_dict[name])
+            try:
+                mod.load_state_dict(saved_state_dict[name])
+            except Exception:
+                if name in saved_state_dict:
+                    logger.warning(f"Failed to load directly for module {name}.")
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/model_saver/torch_model_saver.py:96]:
```suggestion
                    logger.warning(f"Failed to load directly for module {name}, will attempt partial load.")
```

Or something similar

File: ml-agents/mlagents/trainers/model_saver/torch_model_saver.py [hunk 2]
```diff
@@ -89,7 +89,25 @@ def _load_model(
         policy = cast(TorchPolicy, policy)
 
         for name, mod in modules.items():
-            mod.load_state_dict(saved_state_dict[name])
+            try:
+                mod.load_state_dict(saved_state_dict[name])
+            except Exception:
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/model_saver/torch_model_saver.py:94]: What's the specific exception here that we can put? Rather than the generic `Exception`

File: ml-agents/mlagents/trainers/model_saver/torch_model_saver.py [hunk 3]
```diff
@@ -89,7 +89,16 @@ def _load_model(
...
 
         for name, mod in modules.items():
-            mod.load_state_dict(saved_state_dict[name])
+            try:
+                missing_keys, _ = mod.load_state_dict(
+                    saved_state_dict[name], strict=False
+                )
+                if missing_keys:
+                    logger.warning(
+                        f"Did not find these keys {missing_keys} in checkpoint. Initializing"
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/model_saver/torch_model_saver.py:98]:
```suggestion
                        f"Did not find these keys {missing_keys} in checkpoint. Initializing."
```

File: ml-agents/mlagents/trainers/model_saver/torch_model_saver.py [hunk 4]
```diff
@@ -89,7 +89,21 @@ def _load_model(
...
+                    )
+                if unexpected_keys:
+                    logger.warning(
+                        f"Did not expect these keys {unexpected_keys} in checkpoint. Ignoring"
+                    )
+
+            except (KeyError, TypeError):
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/model_saver/torch_model_saver.py:105]: Can you logger.debug the original error ?

Comment (andrewcoh) [reply ml-agents/mlagents/trainers/model_saver/torch_model_saver.py:105]: Instead of or in addition to this warning?

File: ml-agents/mlagents/trainers/model_saver/torch_model_saver.py (hunk 1 above)
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/model_saver/torch_model_saver.py]: This case also catches things that were not present in the previous run e.g. a newly added GAIL head. So it may not have been 'changed'

File: ml-agents/mlagents/trainers/model_saver/torch_model_saver.py [hunk 5]
```diff
@@ -89,7 +89,21 @@ def _load_model(
...
+                logger.warning(f"Failed to load for module {name}. Initializing")
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/model_saver/torch_model_saver.py:106]:
```suggestion
            except Exception as err:
                logger.warning(f"Failed to load for module {name}. Initializing")
                logger.debug(f"Module loading error : {err}")
                
              
```

File: docs/Training-ML-Agents.md [hunk 6]
```diff
@@ -117,6 +117,13 @@ Python by using both the `--resume` and `--inference` flags. Note that if you
 want to run inference in Unity, you should use the
 [Unity Inference Engine](Getting-Started.md#running-a-pre-trained-model).
 
+Additionally, if the network architecture changes, you may still load an existing model,
+and ML-Agents will only load the parts of the model that haven't changed. For instance,
```
Comment (vincentpierre) [inline docs/Training-ML-Agents.md:121]: This is ambiguous because the user does not know which parts would change. Adding a new reward signal requires the critic to be re-initialized for example.

File: ml-agents/mlagents/trainers/model_saver/torch_model_saver.py [hunk 7]
```diff
@@ -89,7 +89,34 @@ def _load_model(
...
+                else:
+                    # optimizers are treated separately
+                    mod.load_state_dict(saved_state_dict[name])
+
+            # KeyError is raised if the module was not present in the last run but is being
+            # accessed in the saved_state_dict.
+            # ValueError is raised by the optimizer's load_state_dict if the parameters have
+            # have changed. Note, the optimizer uses a completely different load_state_dict
+            # function because it is not an nn.Module.
+            # RuntimeError is raised by PyTorch if there  is a size mismatch between modules
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/model_saver/torch_model_saver.py:114]:
```suggestion
            # RuntimeError is raised by PyTorch if there is a size mismatch between modules
```

File: ml-agents/mlagents/trainers/model_saver/torch_model_saver.py [hunk 8]
```diff
@@ -89,7 +89,34 @@ def _load_model(
...
+            # of the same name. This will still partially assign values to those layers that
+            # have no changed shape.
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/model_saver/torch_model_saver.py:116]:
```suggestion
            # have not changed shape.
```

File: ml-agents/mlagents/trainers/tests/torch/saver/test_saver.py [hunk 9]
```diff
@@ -70,6 +72,50 @@ def test_load_save_policy(tmp_path):
...
+    trainer_params2.network_settings = NetworkSettings(
+        hidden_units=10, vis_encode_type=EncoderType(vis_encode_type)
+    )
+    model_saver2 = TorchModelSaver(trainer_params2, path1, load=True)
+    policy2 = create_policy_mock(trainer_params2, use_visual=True)
+    conv_params2 = [mod for mod in policy2.actor.parameters() if len(mod.shape) > 2]
+    # asserts convolutions have different parameters before load
+    for conv1, conv2 in zip(conv_params, conv_params2):
+        assert not torch.equal(conv1, conv2)
+    # asserts layers still  have different dimensions
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/tests/torch/saver/test_saver.py:104]:
```suggestion
    # asserts layers still have different dimensions
```

File: ml-agents/mlagents/trainers/tests/torch/saver/test_saver.py [hunk 10]
```diff
@@ -70,6 +72,50 @@ def test_load_save_policy(tmp_path):
...
+    for mod1, mod2 in zip(policy.actor.parameters(), policy2.actor.parameters()):
+        if mod1.shape[0] == 12:
+            assert mod2.shape[0] == 10
+    model_saver2.register(policy2)
+    model_saver2.initialize_or_load(policy2)
+    # asserts convolutions have same parameters after load
+    for conv1, conv2 in zip(conv_params, conv_params2):
+        assert torch.equal(conv1, conv2)
+    # asserts layers still  have different dimensions
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/tests/torch/saver/test_saver.py:113]:
```suggestion
    # asserts layers still have different dimensions
```

File: docs/Training-ML-Agents.md (hunk 6 above)
Comment (vincentpierre) [reply docs/Training-ML-Agents.md:121]:
```suggestion
but ML-Agents will only load the parts of the model it can load and ignore all others. For instance,
```
This is still ambiguous but the user will not try to interpret what "it can load" means (while they will try to interpret "the parts that haven't changed")

**Review by vincentpierre — APPROVED** [review]

File: ml-agents/mlagents/trainers/model_saver/torch_model_saver.py [hunk 11]
```diff
@@ -89,7 +89,34 @@ def _load_model(
...
+                    if missing_keys:
+                        logger.warning(
+                            f"Did not find these keys {missing_keys} in checkpoint. Initializing."
+                        )
+                    if unexpected_keys:
+                        logger.warning(
+                            f"Did not expect these keys {unexpected_keys} in checkpoint. Ignoring."
+                        )
+                else:
+                    # optimizers are treated separately
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/model_saver/torch_model_saver.py:106]:
```suggestion
                    # If module is not an nn.Module, try to load as one piece
```

**Review by ervteng — APPROVED** [review]
Add to changlog.md. I'd say this qualifies as a "Major change" for the Python package.
